Add Error Tracking + Sentry SDK documentation (#28778)
DataDog/documentation: content/en/error_tracking/sentry_sdk.md

---
title: Sentry SDK

> description: Use the Sentry SDK to send errors to Error Tracking.

Reviewer: ```suggestion
description: Start using Datadog Error Tracking today without changing your existing setup with the Sentry SDK
```

further_reading:
- link: "/error_tracking/manage_data_collection"
  tag: "Documentation"
  text: "Manage Data Collection"
---
{{< callout url="#" btn_hidden="true" >}}
Using the Sentry SDK with Error Tracking is in Preview.
{{< /callout >}}

<div class="alert alert-warning">

> Using the Sentry SDK with Error Tracking helps you migrate to Datadog. However, to get the most out of Error Tracking, it is recommended to use the regular setup. See <a href="/error_tracking/frontend">Frontend Error Tracking</a> and <a href="/error_tracking/backend">Backend Error Tracking</a>.

Reviewer: ```suggestion
Using the Sentry SDK with Error Tracking helps you migrate to Datadog. However, to get the most out of Error Tracking, it is recommended to use the Datadog SDKs. See <a href="/error_tracking/frontend">Frontend Error Tracking</a> and <a href="/error_tracking/backend">Backend Error Tracking</a>.
```

</div>

## Overview
You can use [Sentry SDKs][1] to send your events to Datadog. This offers you a way to start using Error Tracking on existing applications that are monitored using Sentry SDKs.

Setting up the Sentry SDK with Datadog requires a minimal code change to point the SDK to a Datadog Data Source Name (DSN).

[Events][6] are sent to Datadog as logs. Non-error events (messages) also show up as logs. Other item types (traces, attachments, sessions, ...) are not supported.

## Setup
### Prerequisites

> - Sentry SDK events are sent into Datadog as **logs**. You must have [Error Tracking for Logs][2] enabled for errors to show up in Error Tracking.

Reviewer: Maybe we mention.

"To enable Error Tracking for Logs for only your Sentry logs, create a rule (link) like this:" and we provide an example

Author: Good suggestion! added!

### Service configuration
To configure the Sentry SDK to send events into Datadog:
- Configure a Datadog Data Source Name (DSN). Follow the [in-app instructions][5] to generate your unique DSN.
- Set a `service` tag on all events. This is used to separate errors and is shown in the Datadog UI.

{{< tabs >}}

{{% tab "JavaScript" %}}
{{< code-block lang="javascript" >}}
Sentry.init({
    dsn: 'https://myToken@sentry-intake.datadoghq.com/1',
    initialScope: {
        tags: {
            service: 'my-app'
        }
    }
});
{{< /code-block >}}

> {{% /tab }}

Reviewer 2: ```suggestion
{{% /tab %}}
```

Sentry.set_tags('git.repository_url', '<git-provider.example/me/my-repo>')
{{< /code-block >}}
{{% /tab %}}

{{< /tabs >}}

## Supported SDKs

The following Sentry SDKs are verified to work with Error Tracking:

| Platform   | Tested version                                    |
| ---------- | ------------------------------------------------- |
| JavaScript | `@sentry/node@9.13.0`<br>`@sentry/browser@9.13.0` |
| Python     | `sentry-sdk==2.26.1`                              |
| Java       | `io.sentry:sentry:8.6.0`                          |
| .NET       | `Sentry 5.5.1`                                    |
| Go         | `sentry-go v0.32.0`                               |
| Ruby       | `sentry-ruby 5.23.0`                              |

## Migrate to the recommended setup

To get the most out of Error Tracking, it is recommended to eventually migrate to the Datadog SDK and/or Agent-based setups. See [Backend Error Tracking][7] and [Frontend Error Tracking][8] on how to configure this.

The setup using the Sentry SDK can be used simultaneously with the recommended setup. Errors may be reported twice.

## Send events to both Sentry and Datadog

> Events can be sent to both Sentry (or any other Sentry-compatible backend) and Datadog. This allows you to start using Datadog while also keeping your current solution. There are several ways to achieve this.

Reviewer 3: ```suggestion
During your migration, you can use the Sentry SDK and Datadog at the same time.

**Note**: Doing so may result in errors being reported twice.
```

### From the Sentry SDK

> Sentry SDKs can be configured to send events to multiple DSNs at once. On most Sentry SDKs, you can override the default transport to achieve this.

Reviewer 3: ```suggestion
You can configure Sentry SDKs to send events to multiple DSNs at once. On most Sentry SDKs, you can override the default transport to achieve this.
```

{{< tabs >}}

{{% tab "JavaScript" %}}
{{< code-block lang="javascript" >}}
// Change to import from "@sentry/react", "@sentry/nextjs", etc. as needed
import * as Sentry from "@sentry/browser";
import { makeFetchTransport } from "@sentry/browser"; // import { makeNodeTransport } from "@sentry/node" for Node.js
import { makeMultiplexedTransport } from "@sentry/core";

const sentryDsn = '<SENTRY_DSN>';
const datadogDsn = '<DATADOG_DSN>';

Sentry.init({
  dsn: sentryDsn,
  transport: makeMultiplexedTransport(makeFetchTransport, () => [sentryDsn, datadogDsn]),
  // ...
});
Sentry.setTag('service', 'my-app');
{{< /code-block >}}
{{% /tab %}}

{{% tab "Python" %}}

> - Copy the following function into your code:

Reviewer 3: ```suggestion
1. Copy the following function into your code:
```

{{< code-block lang="python" >}}
from sentry_sdk.transport import Transport, make_transport
def make_multi_transport(dsns):
    class MultiTransport(Transport):
        def __init__(self, options):
            super().__init__(options)
            self.transports = [
                make_transport({**options, "dsn": dsn, "transport": None}) for dsn in dsns
            ]
        def capture_envelope(self, *args, **kwargs):
            for transport in self.transports:
                transport.capture_envelope(*args, **kwargs)
        def flush(self, *args, **kwargs):
            for transport in self.transports:
                transport.flush(*args, **kwargs)
        def kill(self):
            for transport in self.transports:
                transport.kill()
        def record_lost_event(self, *args, **kwargs):
            for transport in self.transports:
                transport.record_lost_event(*args, **kwargs)
    return MultiTransport
{{< /code-block >}}

> - Use as follows:

Reviewer 3: ```suggestion
2. Use as follows:
```
I find this phrasing a little awkward and vague, and we use it a few times. I'm wondering if there's more context we can give - for example, why these code snippets are separated out and where they should go?

{{< code-block lang="python" >}}
_SENTRY_DSN = "<SENTRY_DSN>"
_DATADOG_DSN = "<DATADOG_DSN>"
sentry_sdk.init(
    dsn=_SENTRY_DSN,
    transport=make_multi_transport([_SENTRY_DSN, _DATADOG_DSN]),
    # ...
)
sentry_sdk.set_tag("service", "my-app")
{{< /code-block >}}
{{% /tab %}}

{{% tab "Java" %}}

> - Copy the following class into your code. Make sure to define it in the **io.sentry** package.

Reviewer 3: ```suggestion
1. Copy the following class into your code. Make sure to define it in the `io.sentry` package.
```

          return new AsyncHttpTransportFactory().
              create(options, new RequestDetailsResolver(requestOptions).resolve());
        })
        .toList();
    return new ITransport() {
      @Override
      public void send(SentryEnvelope envelope, Hint hint) throws IOException {
        for (ITransport transport : transports) {
          transport.send(envelope, hint);
        }
      }

      @Override
      public boolean isHealthy() {
        return transports.stream().allMatch(ITransport::isHealthy);
      }

      @Override
      public void flush(long timeoutMillis) {
        transports.forEach(transport -> transport.flush(timeoutMillis));
      }

      @Override
      public RateLimiter getRateLimiter() {
        return null;
      }

      @Override
      public void close(boolean isRestarting) throws IOException {
        for (ITransport transport : transports) {
          transport.close(isRestarting);
        }
      }

      @Override
      public void close() throws IOException {
        for (ITransport transport : transports) {
          transport.close();
        }
      }
    };
  }
}
{{< /code-block >}}

> - Use as follows:

Reviewer 3: ```suggestion
1. Use as follows:
```

{{< code-block lang="java" >}}
final var sentryDsn = "<SENTRY_DSN>"
final var datadogDsn = "<DATADOG_DSN>"

Sentry.init(options -> {
  options.setDsn(sentryDsn);
  options.setTransportFactory(new MultiTransportFactory(List.of(sentryDsn, datadogDsn)));
});
Sentry.setTag("service", "my-app");
{{< /code-block >}}
{{% /tab %}}

{{% tab "Go" %}}

> - Copy the following snippet into your code:

Reviewer 3: ```suggestion
1. Copy the following snippet into your code:
```

{{< code-block lang="go" >}}
type MultiTransport struct {
	dsns       []string
	transports []*sentry.HTTPTransport
}

func NewMultiTransport(dsns []string) *MultiTransport {
	transports := make([]*sentry.HTTPTransport, len(dsns))
	for i := range dsns {
		transports[i] = sentry.NewHTTPTransport()
	}
	return &MultiTransport{
		dsns:       dsns,
		transports: transports,
	}
}

func (mt *MultiTransport) Configure(options sentry.ClientOptions) {
	for i := range mt.dsns {
		options.Dsn = mt.dsns[i]
		if options.EnableTracing {
			// Replicating the default behavior:
			// https://github.com/getsentry/sentry-go/blob/v0.32.0/client.go#L358
			mt.transports[i].BufferSize = 1000
		}
		mt.transports[i].Configure(options)
	}
}

func (mt *MultiTransport) Flush(timeout time.Duration) bool {
	allDone := true
	for _, t := range mt.transports {
		if ok := t.Flush(timeout); !ok {
			allDone = false
		}
	}
	return allDone
}

func (mt *MultiTransport) SendEvent(event *sentry.Event) {
	for _, t := range mt.transports {
		t.SendEvent(event)
	}
}

func (mt *MultiTransport) Close() {
	for _, t := range mt.transports {
		t.Close()
	}
}
{{< /code-block >}}

> - Use as follows:

Reviewer 3: ```suggestion
1. Use as follows:
```
